Add type definitions

demurgos opened the pull request:
This commit add type definitions for `gulp-typedoc`. These type
definitions are more complete than the ones from DT. They export all
the types (`Options` and `Logger`), are documented and tested.

The tests try to compile a file in `noEmit` mode to verify that it
type-checks. This adds the new devDevdependencies `@types/node` and
`typescript`.

This is a semver minor change (feature).

- See rogierschouten/gulp-typedoc#20

demurgos commented:
I prefer to add the type definitions here because it's the recommended approach (host types with the code if possible) and hope that it will help them to stay up-to-date. You also stated your interest in https://github.com/rogierschouten/gulp-typedoc/issues/20#issuecomment-322266343.

rogierschouten commented on package.json:
```diff
@@ -8,7 +8,7 @@
-    "test": "mocha test/test.js"
+    "test": "mocha test/test.js && tsc -p test-types"
   },
   "keywords": [
```
Please also add a 'types' field pointing to index.d.ts.  Even though this is the default, according to https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html it is still advisable to have the field in.

```
Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.
```

rogierschouten commented:
Hi, you're back! Thanks for the PR.  Couple of questions:

1. Should @types/node be a normal dependency (or better: peerDependency with broad range)?  Since we publish typings that depend on NodeJS.ReadWriteStream you cannot use the typings without also having the Node typings right? 

1. I added a review comment to add a 'types' field to package.json; even though it's not technically necessary the typescript handbook still advises it

1. The Options interface ties this gulp plugin to a specific version of typedoc.  What would be better is to import the typedoc typings and use their definition. That way, depending on which typedoc typings you install, you get the corresponding options.  HOWEVER, the problem is that the typedoc .d.ts files currently define the Options interface as 'Object'. Bummer.  And you really want the Options interface because that's what makes these typings useful.  Would you consider changing the typedoc typings first and then referencing them?  If that doesn't work I'd be happy with your current solution too.

demurgos commented:
@rogierschouten 
I updated my PR. You're right: the `types` field should be there but I forgot to add it. Good catch.

Regarding the kind of dependency for `@types/node`, I keep following the guidelines used by _typings_ when we moved to `@types`: declarations modifying the global environment should only be installed by the consumer (hence only `devDependencies`). Having normal (or peer) dependencies caused us (@types) some issues at the beginning when we got the first major Node update. Some packages were using `@types/node@7` while others were using `@types/node@8`: theyr interfered and broke many builds. Nowadays it seems that `@types` just marks them as normal dependencies but I guess that it's mostly because their CI and mono-repo structure allows them to push ecosystem-wide dependency updates. Peer-dependencies look good on paper but they are are poor solution: you still have to deal with version issues if dependencies rely on different ranges of `@node/types` and it does not align with Node's model of local dependencies.
I can change it to a normal dependency if you want.
(Most of this info dates from last year and personal experience by maintaining types for typings, I haven't extensively researched if there are some new or official best practices regarding environment declarations)

I totally agree that the Typedoc options should not be declared here but in Typedoc. [I opened a Typedoc issue](https://github.com/TypeStrong/typedoc/issues/677) a few days ago to fix this. Development slowed down so I am not confident in sending a PR at the moment. I'll have more time in a few weeks to push through this and ensure that they refactor their interfaces to be able to expose them.

rogierschouten commented:
@demurgos 

You're right it should not be a dependency.  I guess a peer dependency would in theory be correct - after all, your typings are compatible with a certain range of node typings (currently: all node typings) and you can specify that in the peerDependency field.  But for now I'm happy with them staying a devDependency.

So what do you propose, delay merging this PR until we know if the typedoc guys respond, or merging it now and change it again later?

demurgos commented:
I'd recommend to merge it now and update it once Typedoc is updated. I think that I'll end up sending them a PR but it may take some time. I'll notify you then.

rogierschouten commented:
Merged and published as v2.2.0. Thanks @demurgos!   Maybe also remove typings from DefinitelyTyped now

demurgos commented:
Thanks.
I'll open an issue/PR with DT.
